man/faq/developer/theory-coercion.Rmd

```diff
@@ -0,0 +1,141 @@
+
+Let's look again at what happens when we combine a factor and a character:
+
+```{r}
+vec_ptype2(factor("a"), "b")
+```
+
+The ptype2 method for `<character>` + `<factor>` returns `<character>`. This is because character is a richer type than `factor<"a">`. Whereas the latter can only contain `"a"` strings, the former can contain any strings. Factors are a _subset_ of character.
+
+Note that another valid behaviour would be to throw an incompatible type error. This is what a strict factor implementation would do. We have decided to be laxer in vctrs because it is easy to inadvertently create factors instead of character vectors, especially with older versions of R where `stringsAsFactors` is still true by default.
+
+
+### Consistency and symmetry on permutation
+
+Each ptype2 method should strive to have exactly the same behaviour when the inputs are permuted. This is not always possible, for example factor levels are aggregated in order:
+
+```{r}
+vec_ptype2(factor(c("a", "c")), factor("b"))
+
+vec_ptype2(factor("b"), factor(c("a", "c")))
+```
+
+In any case, permuting the input should not return a fundamentally different type or introduce an incompatible type error.
+
+
+### Coercion hierarchy
+
+Coercible classes form a coercion (or subtyping) hierarchy. Here is a simplified diagram of the hierarchy for base types. In this diagram the directions of the arrows express which type is richer. They flow from the bottom (more constrained types) to the top (richer types).
+
+\\figure{coerce.png}
+
+As a class implementor, you have two options. The simplest is to create an entirely separate hierarchy. The date and date-time classes are an example of an S3-based hierarchy that is completely separate. Alternatively, you can integrate your class in an existing hierarchy, typically by adding parent nodes on top of the hierarchy (your class is richer), by adding children node at the root of the hierarchy (your class is more constrained), or by inserting a node in the tree.
+
+These coercion hierarchies are _implicit_, in the sense that they are implied by the `vec_ptype2()` implementations. There is no structured way to create or modify a hierarchy, instead you need to implement the appropriate coercion methods for all the types in your hierarchy, and diligently return the richer type in each case. The `vec_ptype2()` implementations are not transitive nor inherited, so all pairwise methods between classes lying on a given path must be implemented manually. This is something we might make easier in the future.
+
+
+## `vec_cast()`
+
+The second generic, `vec_cast()`, is the one that look at the data and actually performs the conversion. Because it has access to more information than `vec_ptype2()`, it may be stricter and cause an error in more cases. `vec_cast()` has three possible behaviours:
```

Comment:
```suggestion
The second generic, `vec_cast()`, is the one that looks at the data and actually performs the conversion. Because it has access to more information than `vec_ptype2()`, it may be stricter and cause an error in more cases. `vec_cast()` has three possible behaviours:
```